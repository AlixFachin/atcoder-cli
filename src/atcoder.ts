import {Session} from "./session";
import {ATCODER_BASE_URL, ATCODER_LOGIN_PATH} from "./definitions";
import getConfig from "./config";
import Conf from "conf";
import request, {CookieJar} from "request";

export interface Task {
	id: string,
	label: string,
	title: string,
	url: string
}

export interface Contest {
	id: string,
	title: string,
	url: string
}

export class AtCoder {
	static get base_url(): string {
		return ATCODER_BASE_URL;
	}

	static get login_url(): string {
		return `${AtCoder.base_url}${ATCODER_LOGIN_PATH}`;
	}

	static getContestURL(contest: string) {
		return `${AtCoder.base_url}contests/${contest}`;
	}

	static getTaskURL(contest: string, task: string) {
		return `${AtCoder.getContestURL(contest)}/tasks/${task}`;
	}

	private readonly config: Conf;
	private session: Session;
	// null:未検査 true/false: ログインしているかどうか
	private _login: boolean | null;

	constructor() {
		this.config = getConfig();
		this.session = new Session(this.loadCookiesFromConfig());
		this._login = null;
	}

	/**
	 * ログインしているか調べる
	 * @param force default=false trueならキャッシュを使わずちゃんと調べる
	 */
	async checkSession(force: boolean = false): Promise<boolean> {
		// 以前取得済みならいちいち接続して確かめない
		if (this._login !== null && !force) return this._login;
		return this._login = await this.check();
	}

	/**
	 * アクセスしてログインしている状態かどうかを取得する(結果をキャッシュしない)
	 */
	private async check(): Promise<boolean> {
		// practice contestでログインせず提出ページにアクセスするとコンテストトップに飛ばされることを利用する
		const uri = `${AtCoder.base_url}contests/practice/submit`;
		const response = await this.session.fetch(uri);

		// console.log(response.request.uri.href);
		// リダイレクトされなければログインしている
		return response.request.uri.href === uri;
	}

	/**
	 * ログイン処理します
	 * あまりパスワード文字列を引き回したくないので、この中で標準入力からユーザー名とパスワードを尋ねる
	 */
	async login(): Promise<boolean> {
		if (await this.checkSession()) {
			console.error("you logged-in already");
			return true;
		}
		const csrf_token = await this.getCSRFToken();

		// ユーザーネームとパスワードを入力させる
		const inquirer = await import("inquirer");
		const {username, password} = await inquirer.prompt([{
			type: "input",
			message: "username:",
			name: "username"
		}, {
			type: "password",
			message: "password:",
			name: "password"
		}]) as { username: string, password: string };

		const options = {
			method: "POST",
			formData: {
				username,
				password,
				csrf_token
			}
		};
		const response = await this.session.fetch(AtCoder.login_url, options);
		// ログインページ以外にリダイレクトされていればログイン成功とみなす
		const result = response.request.uri.href !== AtCoder.login_url && response.request.uri.href.indexOf(AtCoder.base_url) === 0;
		if (result) {
			// ログインに成功していた場合はセッション情報を保存する
			this.exportCookiesToConfig(this.session.jar);
		}
		return result;
	}

	/**
	 * ログインページにアクセスしてCSRFトークンを取得
	 */
	private async getCSRFToken(): Promise<string> {
		const {JSDOM} = await import("jsdom");
		const response = await this.session.fetch(AtCoder.login_url);

		const {document} = new JSDOM(response.body).window;
		const input: HTMLInputElement = (document.getElementsByName("csrf_token")[0]) as HTMLInputElement;
		return input.value;
	}

	/**
	 * コンテストIDからコンテストの情報を取得
	 * @param id
	 */
	async contest(id: string): Promise<Contest> {
		const url = AtCoder.getContestURL(id);
		const response = await this.session.fetch(url);
		const {JSDOM} = await import("jsdom");
		const {document} = new JSDOM(response.body).window;
		const regexp = /^(.*) - AtCoder$/;
		const title = regexp.test(document.title) ? regexp.exec(document.title)![1] : document.title;
		return {id, title, url};
	}

	/**
	 * 問題一覧を取得
	 * @param contest
	 */
	async tasks(contest: string): Promise<Array<Task>> {
		const response = await this.session.fetch(`${AtCoder.getContestURL(contest)}/tasks`);

		const {JSDOM} = await import("jsdom");
		const {document} = new JSDOM(response.body).window;
		// very very ad-hoc and not type-safe section
		const tbody = document.querySelector("#main-div .row table>tbody");
		if (tbody === null) return [];
		const tasks: Array<Task> = [];
		for (const tr of tbody.querySelectorAll("tr")) {
			// tr>td>a
			const id: string = tr.children[0].children[0].getAttribute("href")!.split("/").pop()!;
			const label: string = (tr.children[0].children[0] as HTMLAnchorElement).text;
			const title: string = (tr.children[1].children[0] as HTMLAnchorElement).text;
			const url: string = `${AtCoder.base_url}${tr.children[0].children[0].getAttribute("href")!.substring(1)}`;
			tasks.push({id, label, title, url});
		}
		return tasks;
	}

	/**
	 * Configに保存されたCookie情報を読み込む
	 * 引数として与えられたCookieJarの内部状態を書き換えるので、その場合戻り値を再代入する必要はない
	 * @param jar 省略された場合は新しいCookieJarを作って返す
	 */
	private loadCookiesFromConfig(jar?: CookieJar): CookieJar {
		if (jar === undefined) jar = request.jar();
		// configからクッキー情報を取得
		const cookies: string = this.config.get("cookies");
		if (cookies !== undefined) {
			for (const cookie of cookies.split(";")) {
				jar.setCookie(cookie, AtCoder.base_url);
			}
		}
		return jar;
	}

	private exportCookiesToConfig(jar: CookieJar) {
		const cookies = jar.getCookieString(AtCoder.base_url);
		this.config.set("cookies", cookies);
	}
}