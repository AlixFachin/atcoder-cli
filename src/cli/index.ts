import commander from "commander";
// import {version} from "../../package.json";
// work-around to avoid TS5505 error bug in TypeScript 3.1 https://github.com/Microsoft/TypeScript/issues/24715
const {version} = require("../../package.json");
import * as commands from "../commands";

commander
	.version(version, "-v, --version");

commander
	.command("new <contest-id>")
	.alias("setup")
	.action(commands.setup)
	.description("create new contest project directory");

commander
	.command("login")
	.action(commands.login)
	.description("login to AtCoder");

commander
	.command("logout")
	.action(commands.logout)
	.description("delete login session information");

commander
	.command("session")
	.action(commands.session)
	.description("check login or not");

commander
	.command("contest [contest-id]")
	.action(commands.contest)
	.option("-i, --id", "show contest id")
	.description("get contest information");

commander
	.command("tasks [contest-id]")
	.action(commands.tasks)
	.option("-i, --id", "show task id")
	.description("get tasks");

commander
	.command("url [contest] [task]")
	.option("-c, --check", "check the specified contest and/or task id is valid")
	.action(commands.url)
	.description("get contest or task URL");

commander
	.command("oj")
	.action(commands.oj)
	.description("call online-judge");

commander.parse(process.argv);
