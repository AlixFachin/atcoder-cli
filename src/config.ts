import path from "path";

type Conf = import("conf");

export const defaults = {
	["oj-path"]: undefined,
	["default-contest-dirname-format"]: "{ContestID}",
	["default-task-dirname-format"]: "{tasklabel}"
};

/**
 * 常に単一のConfインスタンスを返す関数
 */
const getConfig: () => Promise<Conf> = (() => {
	// confはsingletonとして扱う
	let conf: Conf | null = null;
	return async () => {
		if (conf !== null) return conf;
		const Conf = (await import("conf")).default;
		return conf = new Conf({defaults});
	}
})();

export default getConfig;

/**
 * コンフィグディレクトリを取得
 */
export async function getConfigDirectory(): Promise<string> {
	return path.resolve((await getConfig()).path, "..");
}
