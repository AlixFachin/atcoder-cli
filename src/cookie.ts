import {name as projectName} from "../package.json";
// yummy!
type CookieConf = import("conf")<Array<string>>;

export interface CookieInterface {
	/**
	 * 現在保持しているcookieを取得
	 */
	get(): Array<string>;
	/**
	 * 現在保持しているcookieを新しいものに置き換える
	 * @param cookies 新しく保持するcookie全体を表す、"key=value"形式の文字列の配列
	 */
	set(cookies: Array<string>): void;
	/**
	 * 現在保持しているcookieを捨てる
	 */
	empty(): void;
	/**
	 * 設定ファイルからcookieを読み込んで保持する
	 */
	loadConfigFile(): Promise<void>;
	/**
	 * 現在保持しているcookieを設定ファイルに書き出す
	 */
	saveConfigFile(): Promise<void>;
}

export interface CookieConstructorInterface {
	/**
	 * 設定ファイルからcookie情報を読み込み済みのインスタンスを生成
	 */
	createLoadedInstance(): Promise<Cookie>
}

/**
 * Cookie管理用クラス
 * cookieはHTTPリクエストヘッダにおける Cookie: a=b; c=d
 * の"a=b", "c=d"の部分を文字列の配列として管理する
 */
export class Cookie implements CookieInterface {
	private cookies: Array<string>;
	private static _cookie_conf: CookieConf | null = null;

	protected static async getCookieConfig(): Promise<CookieConf> {
		if (Cookie._cookie_conf !== null) return Cookie._cookie_conf;
		const Conf = (await import("conf")).default;
		return Cookie._cookie_conf = new Conf({defaults: {cookies: []}, configName: "session", projectName});
	}

	/**
	 * 設定ファイルからcookie情報を読み込み済みのインスタンスを生成
	 */
	static async createLoadedInstance(): Promise<Cookie> {
		const cookie = new Cookie();
		await cookie.loadConfigFile();
		return cookie;
	}

	constructor() {
		this.cookies = [];
	}

	/**
	 * 現在保持しているcookieを取得
	 */
	get(): Array<string> {
		return this.cookies;
	}

	/**
	 * 現在保持しているcookieを新しいものに置き換える
	 * @param cookies 新しく保持するcookie全体を表す、"key=value"形式の文字列の配列
	 */
	set(cookies: Array<string>) {
		this.cookies = cookies;
	}

	/**
	 * Set-Cookieヘッダの配列からこのクラスで管理可能な形式のcookie配列に変換
	 * @param cookies
	 */
	static convertSetCookies2CookieArray(cookies: Array<string>): Array<string> {
		return cookies.map(cookie => /^\s*(?:Set-Cookie:\s*)?(.*)$/i.exec(cookie)![1].trim().split(";")[0])
	}

	/**
	 * 現在保持しているcookieを捨てる
	 */
	empty(): void {
		this.cookies = [];
	}

	/**
	 * 設定ファイルからcookieを読み込んで保持する
	 */
	async loadConfigFile(): Promise<void> {
		this.cookies = (await Cookie.getCookieConfig()).get("cookies");
	}

	/**
	 * 現在保持しているcookieを設定ファイルに書き出す
	 */
	async saveConfigFile(): Promise<void> {
		(await Cookie.getCookieConfig()).set("cookies", this.cookies);
	}
}
