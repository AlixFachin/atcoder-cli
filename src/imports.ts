export async function importFsExtra(): Promise<typeof import("fs-extra")> {
	return await import("fs-extra");
}
