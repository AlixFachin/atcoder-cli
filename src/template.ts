import {promisify} from "util";
import {readFile, readdir} from "fs";
import {resolve} from "path";

export const TEMPLATE_JSON_FILE_NAME = "template.json";

export interface Template {
	submit: string;
	program: Array<string>;
	static?: Array<string>;
	cmd?: string;
	testdir?: string;
}

/**
 * ディレクトリ全体を走査してテンプレートディレクトリを取得する
 * @param path
 */
export async function getTemplates(path: string): Promise<Array<{ name: string, template: Template }>> {
	const files: Array<string> = await promisify(readdir)(path);
	const templates = [];
	for (const name of files) {
		try {
			// ファイル名をディレクトリ名とみなして直下のJSONファイルを取得
			const template = JSON.parse(await promisify(readFile)(resolve(path, name, TEMPLATE_JSON_FILE_NAME), "utf8"));
			const [valid, error] = await validateTemplateJSON(template);
			if (valid) {
				templates.push({name, template});
			}
			else {
				console.error(`validation error in ${resolve(name, TEMPLATE_JSON_FILE_NAME)}:\n  ${error}`);
			}
		} catch (e) {
			// ディレクトリでないファイルなど、ENOTDIR, ENOENTエラーが発生した場合はそのまま無視
			if (e.code !== "ENOTDIR" && e.code !== "ENOENT") console.error(`Error occurred in ${resolve(name, TEMPLATE_JSON_FILE_NAME)}:\n  ${e.toString()}`);
		}
	}
	return templates;
}

/**
 * プロジェクトファイルが正しい形式に沿っているか調べる
 * [valid:true, error:null] | [valid:false, error:string] valid=trueなら正しいJSON
 * @param data
 */
export async function validateTemplateJSON(data: Template): Promise<[true, null] | [false, string]> {
	const ajv = new ((await import("ajv")).default)();
	const validate = ajv.compile((await import("./schema")).template_schema);
	const valid = validate(data);
	if (!valid) return [false, ajv.errorsText(validate.errors)];
	return [true, null];
}
