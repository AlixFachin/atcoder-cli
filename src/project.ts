import {Task, ContestProject, Contest} from "./definitions";
import {AtCoder} from "./atcoder";
import {mkdir, readFile, writeFile} from "fs";
import {sep, resolve} from "path";
import mkdirp from "mkdirp";
import {promisify} from "util";
import {OnlineJudge} from "./facade/oj";

export const PROJECT_JSON_FILE_NAME = "contest.acc.json";

/**
 * 指定したディレクトリから親を辿って最も近い位置にあるプロジェクトファイルを取得する
 * 見つからなかった場合は例外を発生させる
 * @param path 省略するとカレントディレクトリを使用
 */
export const findProjectJSON = async (path?: string): Promise<{ path: string, data: ContestProject }> => {
	const readFilePromise = promisify(readFile);
	let cwd = path !== undefined ? path : process.cwd();

	let data = null;
	while (true) {
		try {
			let filepath = resolve(cwd, PROJECT_JSON_FILE_NAME);
			data = JSON.parse(await readFilePromise(filepath, "utf8"));
			break;
		} catch (e) {
			if (e.code === "ENOENT") {
				// ファイルが存在しないので上の階層を探す
				const parent = resolve(cwd, "..");
				if (parent === cwd) {
					throw new Error(`${PROJECT_JSON_FILE_NAME} not found.`);
				}
				cwd = parent;
			}
			else {
				throw e;
			}
		}
	}
	const [valid, error] = await validateProjectJSON(data);
	if (valid) return {path: cwd, data};
	else throw new Error(`failed to validate JSON: ${error!}`);
};

/**
 * プロジェクトファイルを探し、現在のディレクトリ構造からコンテストと問題を特定する
 * @param path? 省略するとカレントディレクトリを使用
 */
export const detectTaskByPath = async (path?: string): Promise<{ contest: Contest | null, task: Task | null }> => {
	if (path === undefined) path = process.cwd();
	try {
		const {path: project_path, data: {contest, tasks}} = await findProjectJSON();
		if (path === project_path) {
			// ブロジェクトディレクトリとカレントディレクトリが一致
			return {contest, task: null};
		}
		// projectディレクトリの一つ下の階層のディレクトリ名を取得
		const dirname = path.split(sep)[project_path.split(sep).length];
		let task = null;
		for (const t of tasks) {
			if (t.id === dirname) {
				task = t;
				break;
			}
		}
		return {contest, task};
	} catch {
		return {contest: null, task: null};
	}
};

/**
 * プロジェクトファイルが正しい形式に沿っているか調べます
 * [valid:true, error:null] | [valid:false, error:string] valid=trueなら正しいJSON
 * @param data
 */
export const validateProjectJSON = async (data: string): Promise<[true, null] | [false, string]> => {
	const ajv = new ((await import("ajv")).default)();
	const schema = (await import("./schema")).default;
	const validate = ajv.compile(schema);
	const valid = validate(data);
	if (!valid) return [false, ajv.errorsText(validate.errors)];
	return [true, null];
};

/**
 * コンテスト情報を取得し、プロジェクトディレクトリを作成する
 * @param contest_id
 */
export const init = async (contest_id: string): Promise<ContestProject> => {
	const atcoder = new AtCoder();
	if (!await atcoder.checkSession()) await atcoder.login();
	const [contest, tasks] = await Promise.all([atcoder.contest(contest_id), atcoder.tasks(contest_id)]).catch(() => [null, null]);
	if (contest === null || tasks === null) {
		throw new Error("failed to get contest information.");
	}
	try {
		await promisify(mkdir)(contest_id);
	}
	catch {
		throw new Error(`${contest_id} file/directory already exists.`)
	}
	process.chdir(contest_id);
	const data = {contest, tasks};
	await promisify(writeFile)(PROJECT_JSON_FILE_NAME, JSON.stringify(data, undefined, 2));
	console.log(`${contest_id}/${PROJECT_JSON_FILE_NAME} created.`);
	return data;
};

export const installTask = async (task: Task, project_path?: string): Promise<void> => {
	const pwd = process.cwd();
	if (project_path !== undefined) {
		await promisify(mkdirp)(project_path);
		process.chdir(project_path);
	}
	await promisify(mkdirp)(task.id);
	process.chdir(task.id);
	if (OnlineJudge.checkAvailable()) {
		await OnlineJudge.call(["dl", task.url, "-d", "tests"]);
	} else {
		console.error("online-judge-tools is not available. downloading of sample cases skipped.");
	}
	// もとのディレクトリに戻る
	process.chdir(pwd);
};
