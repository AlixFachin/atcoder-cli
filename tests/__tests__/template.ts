import * as template from "../../src/template";
import {Contest, Task} from "../../src/project";
import {AtCoder} from "../../src/atcoder";
import {getConfigDirectory} from "../../src/config";
import {importFsExtra} from "../../src/imports";
import {resolve} from "path";
import {promisify} from "util";
import {readFile, readdir} from "fs";
import fs_extra from "fs-extra";
import mock from "mock-fs";
import child_process from "child_process";

jest.mock("../../src/config");
jest.mock("../../src/imports");

const mock_template_cpp: template.RawTemplate = {
	contest: {
		static: [["gitignore", ".gitignore"]]
	},
	task: {
		submit: "main.cpp",
		program: ["main.cpp"]
	}
};
const mock_template_cpp_json = JSON.stringify(mock_template_cpp);
const mock_template_ts: template.RawTemplate = {
	contest: {
		static: ["package.json", ["gitignore", ".gitignore"]],
		cmd: "npm install"
	},
	task: {
		submit: "{TaskID}.ts",
		program: [["main.ts", "{TaskID}.ts"]],
		static: ["foo.txt", ["bar.txt", "{alphabet}_{TaskID}"], ["baz.txt", "{CONTESTID}-{index1}.txt"]],
		cmd: "echo $TASK_ID",
		testdir: "tests-{TaskLabel}"
	}
};
const mock_template_ts_json = JSON.stringify(mock_template_ts);
const mock_templates: Array<{ name: string, template: template.RawTemplate }> = [
	{
		name: "cpp",
		template: mock_template_cpp
	},
	{
		name: "ts",
		template: mock_template_ts
	}
];
const templates2files = (templates: Array<{ name: string, template: template.RawTemplate }>) => templates.reduce(
	(o, {name, template}) => ({
		...o,
		[name]: {"template.json": JSON.stringify(template)}
	}), {});

const DUMMY_CONFIG_DIRECTORY_PATH = "/config/dir/of/acc";
const DUMMY_WORKING_DIRECTORY_PATH = "/current/working/directory/to/solve/problems";

// TODO: もっと適当な場所に定義を移す
const dummy_contest: Contest = {
	id: "aic987",
	title: "AtCoder Imaginary Contest 987",
	url: AtCoder.getContestURL("aic987")
};

const dummy_task01: Task = {
	id: "aic987_a",
	label: "A",
	title: "This is Problem",
	url: AtCoder.getTaskURL("aic987", "aic987_a")
};

describe("template", () => {
	beforeAll(async () => {
		// 常に同じpathを返すようにしておく
		// @ts-ignore: dynamically added method for test
		getConfigDirectory.mockResolvedValue(DUMMY_CONFIG_DIRECTORY_PATH);

		// dynamic import in a mocked filesystem causes errors
		// @ts-ignore: dynamically added method for test
		importFsExtra.mockResolvedValue(fs_extra);
	});
	beforeEach(() => {
		// clear mock call log
		jest.clearAllMocks();
	});
	describe("validateTemplateJSON", () => {
		test("valid", async () => {
			const [valid, error] = await template.validateTemplateJSON(JSON.parse(mock_template_cpp_json));
			expect(valid).toBe(true);
			expect(error).toBe(null);
		});
		test("invalid", async () => {
			// TODO: そもそもinvalidなものはany型でなければ入れられないのでおかしい
			const data: any = {task: {}};
			const [valid, error] = await template.validateTemplateJSON(data);
			expect(valid).toBe(false);
			expect(error).toMatchSnapshot();
		});
	});
	describe("getTemplate", () => {
		test("valid template", async () => {
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: {
					"template01": {
						"template.json": mock_template_cpp_json
					}
				}
			});
			expect(await template.getTemplate("template01")).toEqual({name: "template01", ...mock_template_cpp});
			mock.restore();
		});
		test("invalid template", async () => {
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: {
					"template02": {
						/* lack of required property "task" */
						"template.json": JSON.stringify({
							contest: {
								cmd: "echo hi"
							}
						})
					}
				}
			});
			await expect(template.getTemplate("template02")).rejects.toThrowError();
			mock.restore();
		})
	});
	describe("getTemplates", () => {
		test("get all templates", async () => {
			const templates = [];
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: templates2files(mock_templates)
			});
			const result = await template.getTemplates();
			// restore mock before toMatchSnapshot()
			// Otherwise, jest cannot detect the exist snapshot and the test always passes
			mock.restore();
			expect(result).toMatchSnapshot();
		});
		test("no template.json in directory", async () => {
			const templates = [];
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: {
					...templates2files(mock_templates),
					// add directory without template.json
					"non-template": {
						"this-is-not-template.json": JSON.stringify({
							task: {
								submit: "main.cpp",
								program: ["main.cpp"]
							}
						})
					}
				}
			});
			const result = await template.getTemplates();
			mock.restore();
			// there's no template "non-template"
			expect(result).toMatchSnapshot();
		});
		test("files to be ignored", async () => {
			// mock console.error to avoid https://github.com/tschaub/mock-fs/issues/234
			const spy = jest.spyOn(console, "error").mockImplementation(() => null);
			const templates = [];
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: {
					...templates2files(mock_templates),
					// unrelated files in config directory
					"foo.txt": "I am a text file, unrelated to templates!",
					"config.json": JSON.stringify({
						"default-task-choice": "next"
					})
				}
			});

			const result = await template.getTemplates();
			mock.restore();
			// there's no template "non-template"
			expect(result).toMatchSnapshot();
			spy.mockRestore();
		});
		test("invalid template", async () => {
			const templates = [];
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: {
					...templates2files(mock_templates),
					"invalid": {
						// add invalid template
						"template.json": JSON.stringify({
							contest: {
								cmd: "echo ho"
							}
						})
					}
				}
			});
			// NOTE: そもそもここでエラーを出すのは正しいのか？ (警告メッセージを出すなどする？)
			await expect(template.getTemplates()).rejects.toThrowError();
			mock.restore();
		});
	});
	describe("installContestTemplate", () => {
		test("overwrite static files", async () => {
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: {
					"template01": {
						"template.json": mock_template_cpp_json,
						"gitignore": ".git/"
					}
				},
				[DUMMY_WORKING_DIRECTORY_PATH]: {
					[dummy_contest.id]: {
						".gitignore": "this .gitignore will be overwritten by installing of the template"
					}
				}
			});
			// cd to (mocked) current directory
			process.chdir(DUMMY_WORKING_DIRECTORY_PATH);
			expect(process.cwd()).toEqual(DUMMY_WORKING_DIRECTORY_PATH);

			const template01 = await template.getTemplate("template01");
			expect(await template01).toEqual({name: "template01", ...mock_template_cpp});

			const contest_dir_path = resolve(DUMMY_WORKING_DIRECTORY_PATH, dummy_contest.id);
			expect(await promisify(readdir)(contest_dir_path)).toEqual([".gitignore"]);
			expect(await promisify(readFile)(resolve(contest_dir_path, ".gitignore"), "utf-8")).toEqual("this .gitignore will be overwritten by installing of the template");
			await template.installContestTemplate(dummy_contest, template01, contest_dir_path, false);
			expect(await promisify(readdir)(contest_dir_path)).toEqual([".gitignore"]);
			expect(await promisify(readFile)(resolve(contest_dir_path, ".gitignore"), "utf-8")).toEqual(".git/");

			expect(process.cwd()).toEqual(DUMMY_WORKING_DIRECTORY_PATH);

			mock.restore();
		});
		test("exec command", async () => {
			// mock console.error to avoid https://github.com/tschaub/mock-fs/issues/234
			const spy_console_log = jest.spyOn(console, "log").mockImplementation(() => null);
			const spy_console_error = jest.spyOn(console, "error").mockImplementation(() => null);

			const spy_exec = jest.spyOn(child_process, "exec");
			mock({
				[DUMMY_WORKING_DIRECTORY_PATH]: {
					[dummy_contest.id]: {}
				}
			});

			// cd to (mocked) current directory
			process.chdir(DUMMY_WORKING_DIRECTORY_PATH);
			expect(process.cwd()).toEqual(DUMMY_WORKING_DIRECTORY_PATH);

			const template02 = {name: "template02", contest: {cmd: "echo $CONTEST_ID"}, task: mock_template_cpp.task};
			const contest_dir_path = resolve(DUMMY_WORKING_DIRECTORY_PATH, dummy_contest.id);
			await template.installContestTemplate(dummy_contest, template02, contest_dir_path, true);

			mock.restore();
			expect(spy_exec).toBeCalledTimes(1);
			const test_env = {
				TEMPLATE_DIR: resolve(DUMMY_CONFIG_DIRECTORY_PATH, "template02"),
				CONTEST_DIR: contest_dir_path,
				CONTEST_ID: dummy_contest.id
			};
			expect(spy_exec.mock.calls[0][0]).toEqual("echo $CONTEST_ID");
			expect(spy_exec.mock.calls[0][1]).toEqual({env: expect.objectContaining(test_env)});
			spy_console_log.mockRestore();
			spy_console_error.mockRestore();
			spy_exec.mockRestore();
		});
	});
	describe("installTaskTemplate", () => {
		test("overwrite static files", async () => {
			// mock console.error to avoid https://github.com/tschaub/mock-fs/issues/234
			const spy_console_log = jest.spyOn(console, "log").mockImplementation(() => null);
			const spy_console_error = jest.spyOn(console, "error").mockImplementation(() => null);
			mock({
				[DUMMY_CONFIG_DIRECTORY_PATH]: {
					"template02": {
						"template.json": mock_template_ts_json,
						"gitignore": ".git/",
						"main.ts": "/* do nothing */",
						"foo.txt": "FOO",
						"bar.txt": "BAR",
						"baz.txt": "BAZ"
					}
				},
				[DUMMY_WORKING_DIRECTORY_PATH]: {
					[dummy_contest.id]: {
						[dummy_task01.id]: {
							"foo.txt": "foo",
							"bar.txt": "bar"
						}
					},
					".gitignore": "this .gitignore will be overwritten by installing of the template"
				}
			});
			// cd to (mocked) current directory
			process.chdir(DUMMY_WORKING_DIRECTORY_PATH);
			expect(process.cwd()).toEqual(DUMMY_WORKING_DIRECTORY_PATH);

			const template02 = await template.getTemplate("template02");
			expect(await template02).toEqual({name: "template02", ...mock_template_ts});

			const contest_dir_path = resolve(DUMMY_WORKING_DIRECTORY_PATH, dummy_contest.id);
			const task_dir_path = resolve(DUMMY_WORKING_DIRECTORY_PATH, dummy_contest.id, dummy_task01.id);
			expect((await promisify(readdir)(task_dir_path)).sort()).toEqual(["foo.txt", "bar.txt"].sort());
			expect(await promisify(readFile)(resolve(task_dir_path, "foo.txt"), "utf-8")).toEqual("foo");
			expect(await promisify(readFile)(resolve(task_dir_path, "bar.txt"), "utf-8")).toEqual("bar");
			await template.installTaskTemplate({task: dummy_task01, contest: dummy_contest, index: 0, template: template02}, {contest: contest_dir_path, task: task_dir_path});
			expect((await promisify(readdir)(task_dir_path)).sort()).toEqual(["aic987_a.ts", "foo.txt", "bar.txt", "a_aic987_a", "AIC987-1.txt"].sort());
			expect(await promisify(readFile)(resolve(task_dir_path, "aic987_a.ts"), "utf-8")).toEqual("/* do nothing */");
			expect(await promisify(readFile)(resolve(task_dir_path, "foo.txt"), "utf-8")).toEqual("FOO");
			expect(await promisify(readFile)(resolve(task_dir_path, "bar.txt"), "utf-8")).toEqual("bar");
			expect(await promisify(readFile)(resolve(task_dir_path, "a_aic987_a"), "utf-8")).toEqual("BAR");
			expect(await promisify(readFile)(resolve(task_dir_path, "AIC987-1.txt"), "utf-8")).toEqual("BAZ");

			expect(process.cwd()).toEqual(DUMMY_WORKING_DIRECTORY_PATH);

			mock.restore();
			spy_console_log.mockRestore();
			spy_console_error.mockRestore();
		});
	});
});
