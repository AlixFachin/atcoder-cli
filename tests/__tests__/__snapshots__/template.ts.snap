// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template getTemplates files to be ignored 1`] = `
Array [
  Object {
    "contest": Object {
      "static": Array [
        Array [
          "gitignore",
          ".gitignore",
        ],
      ],
    },
    "name": "cpp",
    "task": Object {
      "program": Array [
        "main.cpp",
      ],
      "submit": "main.cpp",
    },
  },
  Object {
    "contest": Object {
      "cmd": "npm install",
      "static": Array [
        "package.json",
        Array [
          "gitignore",
          ".gitignore",
        ],
      ],
    },
    "name": "ts",
    "task": Object {
      "cmd": "echo $TASK_ID",
      "program": Array [
        Array [
          "main.ts",
          "{TaskID}.ts",
        ],
      ],
      "static": Array [
        "foo.txt",
        Array [
          "bar.txt",
          "{alphabet}_{TaskID}",
        ],
        Array [
          "baz.txt",
          "{CONTESTID}-{index1}.txt",
        ],
      ],
      "submit": "{TaskID}.ts",
      "testdir": "tests-{TaskLabel}",
    },
  },
]
`;

exports[`template getTemplates get all templates 1`] = `
Array [
  Object {
    "contest": Object {
      "static": Array [
        Array [
          "gitignore",
          ".gitignore",
        ],
      ],
    },
    "name": "cpp",
    "task": Object {
      "program": Array [
        "main.cpp",
      ],
      "submit": "main.cpp",
    },
  },
  Object {
    "contest": Object {
      "cmd": "npm install",
      "static": Array [
        "package.json",
        Array [
          "gitignore",
          ".gitignore",
        ],
      ],
    },
    "name": "ts",
    "task": Object {
      "cmd": "echo $TASK_ID",
      "program": Array [
        Array [
          "main.ts",
          "{TaskID}.ts",
        ],
      ],
      "static": Array [
        "foo.txt",
        Array [
          "bar.txt",
          "{alphabet}_{TaskID}",
        ],
        Array [
          "baz.txt",
          "{CONTESTID}-{index1}.txt",
        ],
      ],
      "submit": "{TaskID}.ts",
      "testdir": "tests-{TaskLabel}",
    },
  },
]
`;

exports[`template getTemplates no template.json in directory 1`] = `
Array [
  Object {
    "contest": Object {
      "static": Array [
        Array [
          "gitignore",
          ".gitignore",
        ],
      ],
    },
    "name": "cpp",
    "task": Object {
      "program": Array [
        "main.cpp",
      ],
      "submit": "main.cpp",
    },
  },
  Object {
    "contest": Object {
      "cmd": "npm install",
      "static": Array [
        "package.json",
        Array [
          "gitignore",
          ".gitignore",
        ],
      ],
    },
    "name": "ts",
    "task": Object {
      "cmd": "echo $TASK_ID",
      "program": Array [
        Array [
          "main.ts",
          "{TaskID}.ts",
        ],
      ],
      "static": Array [
        "foo.txt",
        Array [
          "bar.txt",
          "{alphabet}_{TaskID}",
        ],
        Array [
          "baz.txt",
          "{CONTESTID}-{index1}.txt",
        ],
      ],
      "submit": "{TaskID}.ts",
      "testdir": "tests-{TaskLabel}",
    },
  },
]
`;

exports[`template validateTemplateJSON invalid 1`] = `"data.task should have required property 'submit'"`;
